tortoise/state.go

```diff
@@ -161,37 +165,44 @@ func (s *state) Recover() error {
 	return nil
 }
 
-func (s *state) Evict() error {
+func (s *state) Evict(ctx context.Context, windowStart types.LayerID) error {
 	var (
-		batch = s.db.NewBatch()
-		it    = s.db.Find([]byte(namespaceOpinions))
-		b     bytes.Buffer
+		logger = s.log.WithContext(ctx)
+		batch  = s.db.NewBatch()
+		b      bytes.Buffer
 	)
-	defer it.Release()
-	for it.Next() {
-		layer := decodeLayerKey(it.Key())
-		if layer.After(s.LastEvicted) {
-			break
-		}
-		b.WriteString(namespaceGood)
-		offset := 1 + types.LayerIDSize
-		b.Write(it.Key()[offset : offset+types.BlockIDSize])
-		if err := batch.Delete(b.Bytes()); err != nil {
-			return fmt.Errorf("delete from batch: %w", err)
+	for layerToEvict := s.LastEvicted.Add(1); layerToEvict.Before(windowStart); layerToEvict = layerToEvict.Add(1) {
+		logger.With().Debug("evicting layer",
+			layerToEvict,
+			log.Int("blocks", len(s.BlockOpinionsByLayer[layerToEvict])))
+		for blk := range s.BlockOpinionsByLayer[layerToEvict] {
+			if s.GoodBlocksIndex[blk] {
+				encodeGoodBlockKey(&b, blk)
+				if err := batch.Delete(b.Bytes()); err != nil {
+					return fmt.Errorf("delete good block in batch %x: %w", b.Bytes(), err)
+				}
+				b.Reset()
+			}
+			delete(s.GoodBlocksIndex, blk)
+			delete(s.BlockLayer, blk)
+			for blk2, opinion := range s.BlockOpinionsByLayer[layerToEvict][blk] {
+				if !opinion.Flushed {
+					continue
+				}
+				encodeOpinionKey(&b, layerToEvict, blk, blk2)
+				if err := batch.Delete(b.Bytes()); err != nil {
+					return fmt.Errorf("delete opinion in batch %x: %w", b.Bytes(), err)
+				}
+				b.Reset()
+			}
 		}
-		b.Reset()
-		if err := batch.Delete(it.Key()); err != nil {
-			return fmt.Errorf("delete from batch: %w", err)
+		delete(s.BlockOpinionsByLayer, layerToEvict)
+		if err := batch.Write(); err != nil {
+			return fmt.Errorf("write batch for layer %s to db: %w", layerToEvict, err)
 		}
+		batch.Reset()
```

Comment:
what is your guideline for calling `batch.Reset()`? data size?

Reply:
pretty much. it should be more efficient to write everything in a single batch, but writing the whole state after rerun was problematic